### PR TITLE
add gzip read in StorageHDFS

### PR DESCRIPTION
**Pull request by apbodrov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category:
- Improvement

Added the ability to read gzip files in StorageHDFS based on the file name

special thanks to @horoshenkiy

### CODE REVIEW

File: dbms/src/Storages/StorageHDFS.cpp [hunk 1]
```diff
@@ -59,7 +61,15 @@ class HDFSBlockInputStream : public IBlockInputStream
         const Context & context,
         UInt64 max_block_size)
     {
-        std::unique_ptr<ReadBuffer> read_buf = std::make_unique<ReadBufferFromHDFS>(uri);
+        std::unique_ptr<ReadBuffer> read_buf;
+        if (boost::algorithm::ends_with(uri, ".gz")) {
```
Comment (alesapin) [inline dbms/src/Storages/StorageHDFS.cpp]: Better to use https://github.com/clickhouse/ClickHouse/blob/master/dbms/src/Common/StringUtils/StringUtils.h#L34.

Comment (alesapin) [reply dbms/src/Storages/StorageHDFS.cpp]: I'm not sure, that the decision of compression type should be made on the file extension. Maybe better to add an optional parameter to`StorageHDFS` with compression name?

Comment (apbodrov) [reply dbms/src/Storages/StorageHDFS.cpp]: > I'm not sure, that the decision of compression type should be made on the file extension. Maybe better to add an optional parameter to`StorageHDFS` with compression name?

Why not both? :)
I will add an optional parameter, but I also want to keep the option to choose a compression method by file extension.

**Comment by alexey-milovidov** [comment]
We should also implement the same support for `File`, `URL` and `S3` storages.

Determining the compression by file extension is Ok. But we should provide a setting to more fine control over it. This setting should have the following values: `none`, `gz` (`bz2`, `xz` are possible to also implement later) and `auto`. The `auto` is autodetect by extension and it can be made default.

And we should also implement it for data output.

**Comment by apbodrov** [comment]
Closing pull request, will make a new

**Pull request closed without merging**